### PR TITLE
Delete action for Container Repository Images and Vault Secrets

### PR DESCRIPTION
**Pull request by petrovic-d**
Action for deleting Container Repository Images and Vault Secrets.
Since there is no API to instantly delete Vault Secrets, the deletion is scheduled for tomorrow. Also Cloud Asset view now shows if Vault Secret is in state: PENDING_DELETION

### CODE REVIEW

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/actions/DeleteNodeAction.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        lazy = true
+)
+
+@ActionReferences(value = {
+    @ActionReference(path = "Cloud/Oracle/ContainerTag/Actions", position = 250),
+    @ActionReference(path = "Cloud/Oracle/Vault/Secret/Actions", position = 250)
+})
+@NbBundle.Messages({
+    "DeleteNodeAction=Delete item",
+    "MSG_ConfirmDeleteAction=Are you sure that you want to delete {0}",
```
Comment (sdedic) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/actions/DeleteNodeAction.java:59]: Nobody funds (much) internationalization ;) these days  ... but still: please place a comment line above parametrized message. The recommended format is
```
# {n} - description of parameter n
```
It used to help translators to understand what individual parameters represent; they work primarily with .properties only.

File: java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/LspAssetsDecorationProvider.java [hunk 2]
```diff
@@ -101,6 +103,11 @@ public TreeItemData createDecorations(Node n, boolean expanded) {
                 d.addContextValues(CTXVALUE_CAP_REFERENCE_NAME);
                 set = true;
             }
+            
+            if (item instanceof SecretItem) {
```
Comment (sdedic) [inline java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/LspAssetsDecorationProvider.java:107]: @jhorvath these interface-dependent extractions ("if X in Lookup, use put "prefix:${X.property}" into contextValue) seem rather common / useful ... if the usage will grow, we could invent a declarative support for that, would be then far easier to export state to LSP clients.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/vault/SecretNode.java [hunk 3]
```diff
@@ -41,13 +56,81 @@ public SecretNode(SecretItem vault) {
...
+        SimpleDateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
+        return "\nDeletion time: " + dateFormat.format(deletionTime);
     }
 
     public static NodeProvider<SecretItem> createNode() {
         return SecretNode::new;
     }
+    
+    @Override
+    public void update(OCIItem item) {
```
Comment (sdedic) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/vault/SecretNode.java:82]: Make sure `update` does not run in a LSP messaging thread, or other 'must not block' thread ...

**Comment by jhorvath** [comment]
The DeleteNodeAction should not be implemented at all. There is already support for deletion. The context value cap:delete is added when Node.canDestroy() returns true in DefaultDecorationsImpl. Additionally, destroy() is called in LspTreeViewServiceImpl for the delete command, which should appear automatically based on the context value.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/NotificationUtils.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    public static boolean confirmAction(String message) {
+        NotifyDescriptor.Confirmation msg = new NotifyDescriptor.Confirmation(message, NotifyDescriptor.YES_NO_OPTION);
+        Object choice = DialogDisplayer.getDefault().notify(msg);
+        return choice == NotifyDescriptor.YES_OPTION || choice == NotifyDescriptor.OK_OPTION;
+    }
+    
+    public static void showErrorMessage(String message) {
+        StringBuilder sb = new StringBuilder("Error\n");
+        sb.append(message);
+        showMessage(sb.toString());
```
Comment (sdedic) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/NotificationUtils.java:39]: if the msg is intended to be presented as an error, then `new NotifyDescriptor.Message(message, NotifyDescriptor.ERROR_MESSAGE)` should be used -- will affect the presentation.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OCINode.java [hunk 5]
```diff
@@ -111,15 +112,21 @@ public static final List<? extends Action> actionsForPath(String path, Lookup lk
     }
     
     public void refresh() {
-        if (factory != null) {
-            factory.refreshKeys();
-        }
-        update(item);
+        RequestProcessor.getDefault().post(() -> {
```
Comment (sdedic) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OCINode.java:115]: Note: the default RP (see the documentation for getDefault()) has a guard against 'too often scheduled task'. Not sure if "refresh" on a parent node (that contains children) refreshes the entire subtree one by one - that may cause the guard to trigger & print some annoying messages into the log. If the refresh only affects a single node, it's fine.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/vault/SecretNode.java [hunk 6]
```diff
@@ -18,36 +18,162 @@
...
+    
+    private String createShortDescription(String state, Date deletionTime) {
+        if (deletionTime != null) {
+            return Bundle.SecretNodeDesc(this.getItem().getName(), state, getDeletionTimeInfo(deletionTime));
+        }
+        return Bundle.SecretNodeDesc(this.getItem().getName(), state, "");
+    }
+    
+    private String getDeletionTimeInfo(Date deletionTime) {
+        return "\nDeletion time: " + DELETION_TIME_FORMAT.format(deletionTime);
```
Comment (sdedic) [inline enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/vault/SecretNode.java:91]: This decoration ought to be in a bundle. Pls. create 2 bundle keys, one normal, the other for deleted items.

File: enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OCINode.java (hunk 5 above)
Comment (petrovic-d) [reply enterprise/cloud.oracle/src/org/netbeans/modules/cloud/oracle/OCINode.java:115]: Just tested it, refresh() method will be triggered only once, no matter the node hierarchy

**Review by sdedic — APPROVED** [review]

**Review by jhorvath — APPROVED** [review]
Looks good

**Pull request merged**